source/agora/common/EnrollmentManager.d

```diff
@@ -0,0 +1,407 @@
+                                a validator and to generate a siging key
+
+        Returns:
+            an enrollment data
+
+    ***************************************************************************/
+
+    public Enrollment createEnrollment (const ref Hash frozen_utxo_hash) @trusted
+    {
+        this.data = new Enrollment();
+
+        // K, frozen UTXO hash
+        this.data.utxo_key = frozen_utxo_hash;
+
+        // N, cycle length
+        this.data.cycle_length = 1008; // freezing period / 2
+
+        // generate random seed value
+        this.random_seed_src = Scalar.random();
+
+        // X, nth image of random seed
+        this.data.random_seed = hashFull(this.random_seed_src);
+        foreach (i; 0 .. this.data.cycle_length-1)
+            this.data.random_seed = hashFull(this.data.random_seed);
+
+        // R, signature noise
+        this.signature_noise = Pair.random();
+
+        // signature
+        Hash hash = hashFull(*this.data);
+        data.enroll_sig = signHash(this.node_key_pair.secret, signature_noise, hash);
+
+        // save information for enrollment data
+        saveNodeEnrollmentData();
+
+        return *this.data;
+    }
+
+    /***************************************************************************
+
+        Check if the enrollment data exists in the validator set.
+
+        Params:
+            enroll_hash = key for an enrollment data which is hash of frozen UTXO 
+
+        Returns:
+            true if the enrollment manager has the enrollment data
+
+    ***************************************************************************/
+
+    public bool hasEnrollment (const ref Hash enroll_hash) @trusted
+    {
+        auto results = this.db.execute("SELECT key, val FROM validator_set " ~
+            "WHERE key = ?", enroll_hash[]);
+
+        return !results.empty;
+    }
+    
+    /***************************************************************************
+
+        Check if the enrollment data exists in the validator set.
+
+        Params:
+            enroll_hash = key for an enrollment data which is a hash of a frozen 
+                            UTXO
+            enroll = will contain the enrollment data if found
+
+        Returns:
+            Return true if the enrollment data was found
+
+    ***************************************************************************/
+
+    public bool getEnrollment (const ref Hash enroll_hash,
+        out Enrollment enroll) @trusted
+    {
+        auto results = this.db.execute("SELECT key, val FROM validator_set " ~
+            "WHERE key = ?", enroll_hash[]);
+
+        foreach (row; results)
+        {
+            enroll = deserializeFull!Enrollment(row.peek!(ubyte[])(1));
+            return true;
+        }
+
+        return false;
+    }
+
+    // sign enrollment data with schnorr key pairs
+    private static Signature signHash (const ref SecretKey secret_key, 
```

Comment:
This function is clearly not carrying its weight and should be removed, unless I am missing something ?